core/payment/src/processor.rs

```diff
@@ -1,203 +1,288 @@
         }
-        .await;
+    }
 
-        if let Err(e) = result {
-            log::error!("Payment failed: {}", e);
+    async fn driver(
+        &self,
+        platform: &str,
+        address: &str,
+        mode: AccountMode,
+    ) -> Result<String, AccountNotRegistered> {
+        match self
+            .accounts
+            .lock()
+            .await
+            .get(&(platform.to_owned(), address.to_owned()))
+        {
+            Some((driver, reg_mode)) if reg_mode.contains(mode) => Ok(driver.to_owned()),
+            _ => Err(AccountNotRegistered::new(platform, address, mode)),
         }
     }
 
-    pub async fn schedule_payment(&self, msg: SchedulePayment) -> PaymentResult<()> {
-        let document_id = msg.document_id();
-        let amount = PaymentAmount {
-            base_currency_amount: msg.amount.clone(),
-            gas_amount: None,
-        };
-        bus::service(driver::BUS_ID)
+    async fn platform(&self, driver: &str) -> Result<String, DriverNotRegistered> {
+        match self.drivers.lock().await.get(driver) {
+            Some(platform) => Ok(platform.to_owned()),
+            None => Err(DriverNotRegistered::new(driver)),
+        }
+    }
+
+    pub async fn notify_payment(&self, msg: NotifyPayment) -> Result<(), NotifyPaymentError> {
+        let payment_platform = self.platform(&msg.driver).await?;
+        let payer_addr = msg.sender;
+        let payee_addr = msg.recipient;
+
+        let orders = self
+            .db_executor
+            .as_dao::<OrderDao>()
+            .get_many(msg.order_ids, msg.driver)
+            .await?;
+        validate_orders(
+            &orders,
+            &payment_platform,
+            &payer_addr,
+            &payee_addr,
+            &msg.amount,
+        )
+        .await?;
+
+        let mut activity_payments = vec![];
+        let mut agreement_payments = vec![];
+        for order in orders.iter() {
+            let amount = order.amount.clone().into();
+            match (order.activity_id.clone(), order.agreement_id.clone()) {
+                (Some(activity_id), None) => activity_payments.push(ActivityPayment {
+                    activity_id,
+                    amount,
+                    allocation_id: Some(order.allocation_id.clone()),
+                }),
+                (None, Some(agreement_id)) => agreement_payments.push(AgreementPayment {
+                    agreement_id,
+                    amount,
+                    allocation_id: Some(order.allocation_id.clone()),
+                }),
+                _ => return NotifyPaymentError::invalid_order(&order),
+            }
+        }
+
+        // FIXME: This is a hack. Payment orders realized by a single transaction are not guaranteed
+        //        to have the same payer and payee IDs. Fixing this requires a major redesign of the
+        //        data model. Payments can no longer by assigned to a single payer and payee.
```

Comment:
This fix is intentionally left for another PR because it requires substantial amount of work.